### PR TITLE
[`test`] Add Dockerfile 

### PR DESCRIPTION
**Pull request by younesbelkada**
This PR will add a proper dockerfile and a github action to push to docker image everyday - based on https://github.com/huggingface/accelerate/blob/main/.github/workflows/build_docker_images.yml & https://github.com/huggingface/accelerate/blob/main/docker/accelerate-gpu/Dockerfile

As `peft` is closely tied to the latest versions of `transformers` & `accelerate` the docker image will contain the main branch of these two libraries 

Also we can discuss about whether it is necessary to have a workflow that pushes the fresh docker image everyday, or manually clone accelerate & transformers on the VM and just fetch the latest changes

### CODE REVIEW

**Comment by younesbelkada** [comment]
I have created a dedicated VM on GCP and tested the docker images locally, this should be good!

File: .github/workflows/build_docker_images.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      - name: Check out code
+        uses: actions/checkout@v2
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_PASSWORD }}
+
```
Comment (younesbelkada) [inline .github/workflows/build_docker_images.yml:27]: are these secrets all shared through HF repos?

**Review by muellerzr — APPROVED** [review]
LG2M from what I can tell!

cc @ydshieh in case you notice anything 😄

File: .github/workflows/build_docker_images.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  latest-cpu:
+    name: "Latest Peft CPU [dev]"
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      - name: Check out code
+        uses: actions/checkout@v2
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
```
Comment (ydshieh) [inline .github/workflows/build_docker_images.yml:23]: v2 ? maybe newer also exist

File: .github/workflows/build_docker_images.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_PASSWORD }}
+
+      - name: Build and Push CPU
+        uses: docker/build-push-action@v2
```
Comment (ydshieh) [inline .github/workflows/build_docker_images.yml:29]: v3 or higher? I will leave you to decide and check

File: docker/peft-cpu/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    git+https://github.com/huggingface/accelerate \
+    git+https://github.com/huggingface/peft#egg=peft[test]
+
+# Install apt libs
+RUN apt-get update && \
+    apt-get install -y curl git wget && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists*
+
+RUN echo "source activate peft" >> ~/.profile
```
Comment (ydshieh) [inline docker/peft-cpu/Dockerfile:44]: why not do this once (before the installation above)?

File: docker/peft-cpu/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+# Activate the virtualenv
+CMD ["/bin/bash"]
```
Comment (ydshieh) [inline docker/peft-cpu/Dockerfile:47]: In the actual Dockfile(s) used in `transformers`, we don't have this. But you can keep it if you want (I have no idea)

**Review by ydshieh — APPROVED** [review]
I don't know everything regarding all these commands, but left a few comments.

I would strongly recommend to trigger the Actions to make sure everything works before merging. Each time I work on docker or workflow files, they are endless of issues when I  tried to run them 😭

File: docker/peft-cpu/Dockerfile (hunk 4 above)
Comment (muellerzr) [reply docker/peft-cpu/Dockerfile:44]: If you mean the multi-stage approach here, it reduces the overall image size by just keeping the compiled parts of everything else. We do this in Accelerate

Comment (ydshieh) [reply docker/peft-cpu/Dockerfile:44]: No, I mean `source activate peft` twice. (Not a big deal, I am just curious)

Comment (muellerzr) [reply docker/peft-cpu/Dockerfile:44]: Oh, it's ran once. That command sends the source activate to the bash profile so on loading a terminal the `peft` env is active

Comment (ydshieh) [reply docker/peft-cpu/Dockerfile:44]: oh, yes. My bad, sorry to bother 😃

Comment (muellerzr) [reply docker/peft-cpu/Dockerfile:44]: No worries! :)

**Review by pacman100 — APPROVED** [review]
Thank you @younesbelkada for adding the docker support, LGTM! 🤗

File: .github/workflows/build_docker_images.yml (hunk 2 above)
Comment (younesbelkada) [inline .github/workflows/build_docker_images.yml:23]:
```suggestion
        uses: docker/login-action@v2
```

File: .github/workflows/build_docker_images.yml (hunk 3 above)
Comment (younesbelkada) [inline .github/workflows/build_docker_images.yml:29]:
```suggestion
        uses: docker/build-push-action@v4
```

File: .github/workflows/build_docker_images.yml [hunk 6]
```diff
@@ -0,0 +1,54 @@
+name: Build Docker images (scheduled)
+
+on:
+  workflow_dispatch:
```
Comment (younesbelkada) [inline .github/workflows/build_docker_images.yml:4]:
```suggestion
  push:
  workflow_dispatch:
```

File: .github/workflows/build_docker_images.yml [hunk 7]
```diff
@@ -0,0 +1,55 @@
+name: Build Docker images (scheduled)
+
+on:
+  push:
```
Comment (younesbelkada) [inline .github/workflows/build_docker_images.yml:4]:
```suggestion
```

**Comment by younesbelkada** [comment]
Thanks everyone for your valuable comments! 💪

**Comment by ydshieh** [comment]
😭 😭😭 @ydshieh always has 110% accuracy for his predictions.

Screenshot 2023-04-11 104424

**Comment by younesbelkada** [comment]
hahahahah indeed

**Pull request merged**